src/utils/all_reduce.hpp

```diff
@@ -0,0 +1,106 @@
+//========================================================================================
+// (C) (or copyright) 2021. Triad National Security, LLC. All rights reserved.
+//
+// This program was produced under U.S. Government contract 89233218CNA000001 for Los
+// Alamos National Laboratory (LANL), which is operated by Triad National Security, LLC
+// for the U.S. Department of Energy/National Nuclear Security Administration. All rights
+// in the program are reserved by Triad National Security, LLC, and the U.S. Department
+// of Energy/National Nuclear Security Administration. The Government is granted for
+// itself and others acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works, distribute copies to
+// the public, perform publicly and display publicly, and to permit others to do so.
+//========================================================================================
+#ifndef UTILS_ALL_REDUCE_HPP_
+#define UTILS_ALL_REDUCE_HPP_
+
+#include <vector>
+
+#include <config.hpp>
+#include <parthenon_mpi.hpp>
+
+namespace parthenon {
+
+enum class Op { MAX, MIN, SUM, PROD, LAND, BAND, LOR, BOR, LXOR, BXOR, MAXLOC, MINLOC };
+#ifndef MPI_PARALLEL
+#define MPI_
+#endif
+
+template <typename T, int N = 0>
+struct AllReduce {
+  T val;
+#ifdef MPI_PARALLEL
+  MPI_Request req;
+#endif
+  bool active = false;
+
+  template <typename U = T>
+  typename std::enable_if<std::is_same<U, Real>::value, TaskStatus>::type
+  StartReduce(int &only, MPI_Op op) {
+#ifdef MPI_PARALLEL
+    if (only == N) {
+      MPI_Iallreduce(MPI_IN_PLACE, &val, 1, MPI_PARTHENON_REAL, op, MPI_COMM_WORLD, &req);
+    }
+#endif
+    active = true;
+    return TaskStatus::complete;
+  }
```

Comment:
This seems like an odd pattern... Can you not just template on `T` here and then provide the separate implementations in a `*.cpp` file?

Reply:
You don't like my SFINAE???

Yes, you can just use template specialization.  My secret ambition is to have all of parthenon be header only.

The other solution I was thinking about was some kind of map from type to MPI_Datatype.  like `my_map[Real] = MPI_PARTHENON_REAL`.  Obviously that doesn't work as I just wrote it, and in any case still requires that you write out all the types explicitly.  I'll think a bit more about what the right thing to do here is...

Reply:
I like the template solution overall. I think I'd prefer standard template specialization over SFINAE. But obviously it's not a big deal.

This is really a style thing more than anything else.

If you were really evil and we had access to it in the standard, you could return a std::variant. (Don't do that.)

Reply:
ok check out my solution now @Yurlungur.  I like this one a lot better.  I even got to use a `void *` and not feel guilty about it 😄

Reply:
Yeah I like the new solution. Very old-style MPI if I do say so myself.